### PR TITLE
routes: changes required to do browser authentication

### PR DESCRIPTION
**Pull request by shimish2**
whenever we make a request that contains header apart from CORS allowed header, browser sends a preflight request
and in response accept *Access-Control-Allow-Headers*.

preflight request is in form of OPTIONS method, added new http handler func to set headers
and returns HTTP status ok in case of OPTIONS method.

in case of authorization, request contains authorization header
added authorization header in Access-Control-Allow-Headers list

added AllowOrigin field in HTTPConfig this field value is set to Access-Control-Allow-Origin header and will give zot adminstrator to limit incoming request.

Signed-off-by: Shivam Mishra <shimish2@cisco.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: golangcilint.yaml [hunk 1]
```diff
@@ -13,7 +13,7 @@ linters-settings:
   nestif:
     min-complexity: 26
   cyclop:
-    max-complexity: 40
```
Comment (rchincha) [inline golangcilint.yaml:16]: ?

File: pkg/api/config/config.go [hunk 2]
```diff
@@ -64,6 +64,7 @@ type RatelimitConfig struct {
 type HTTPConfig struct {
 	Address          string
 	Port             string
+	AllowOrigin      string // comma separated
```
Comment (rchincha) [inline pkg/api/config/config.go:67]: This is good.

File: pkg/api/controller.go [hunk 3]
```diff
@@ -57,19 +57,29 @@ func NewController(config *config.Config) *Controller {
...
 }
 
-func DefaultHeaders() mux.MiddlewareFunc {
+func (c *Controller) CORSHeaders() mux.MiddlewareFunc {
 	return func(next http.Handler) http.Handler {
 		return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
 			// CORS
-			response.Header().Set("Access-Control-Allow-Origin", "*")
-			response.Header().Set("Access-Control-Allow-Methods", "POST, GET, OPTIONS")
+			c.CORSHandler(response, request)
```
Comment (rchincha) [inline pkg/api/controller.go:64]: Why not inline the method below right here? Only caller right?

Comment (shimish2) [reply pkg/api/controller.go:64]: No, it is also called by engine.Use https://github.com/shimish2/zot/blob/ui-browser-auth/pkg/api/controller.go#L133

File: pkg/api/authn.go [hunk 4]
```diff
@@ -45,6 +45,11 @@ func bearerAuthHandler(ctlr *Controller) mux.MiddlewareFunc {
 
 	return func(next http.Handler) http.Handler {
 		return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
+			if request.Method == http.MethodOptions {
+				response.WriteHeader(http.StatusOK)
```
Comment (rchincha) [inline pkg/api/authn.go:49]: Make this 204 since no content

File: pkg/api/authn.go [hunk 5]
```diff
@@ -72,6 +77,37 @@ func bearerAuthHandler(ctlr *Controller) mux.MiddlewareFunc {
 	}
 }
 
+func processNoAuth(realm string, config *config.Config) mux.MiddlewareFunc {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
+			if request.Method == http.MethodOptions {
+				response.WriteHeader(http.StatusOK)
```
Comment (rchincha) [inline pkg/api/authn.go:84]: same 204 here

File: pkg/api/authn.go [hunk 6]
```diff
@@ -84,28 +120,7 @@ func basicAuthHandler(ctlr *Controller) mux.MiddlewareFunc {
...
-					response.WriteHeader(http.StatusMethodNotAllowed)
-
-					return
-				}
-
-				// Process request
-				next.ServeHTTP(response, request)
-			})
-		}
+		return processNoAuth(realm, ctlr.Config)
```
Comment (rchincha) [inline pkg/api/authn.go:123]: noPasswdAuth()

File: pkg/api/authn.go [hunk 7]
```diff
@@ -177,6 +192,11 @@ func basicAuthHandler(ctlr *Controller) mux.MiddlewareFunc {
 
 	return func(next http.Handler) http.Handler {
 		return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
+			if request.Method == http.MethodOptions {
+				response.WriteHeader(http.StatusOK)
```
Comment (rchincha) [inline pkg/api/authn.go:196]: 204 here also

File: pkg/api/controller.go [hunk 8]
```diff
@@ -57,19 +57,29 @@ func NewController(config *config.Config) *Controller {
...
 
+func (c *Controller) CORSHandler(response http.ResponseWriter, request *http.Request) {
+	// allow origin as specified in config if not accept request from anywhere.
+	if c.Config.HTTP.AllowOrigin == "" {
+		response.Header().Set("Access-Control-Allow-Origin", "*")
+	} else {
+		response.Header().Set("Access-Control-Allow-Origin", c.Config.HTTP.AllowOrigin)
+	}
+
+	response.Header().Set("Access-Control-Allow-Methods", "POST, GET, OPTIONS, HEAD")
```
Comment (rchincha) [inline pkg/api/controller.go:79]: Make OPTIONS the last entry

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**